_posts/2022-07-25-Building-CNF-Using-CNDP-Rust.md

```diff
@@ -0,0 +1,194 @@
+footprint, reliability, security, and fearless concurrency at scale
+makes it a good candidate for building Cloud native Network Functions
+(CNFs).
+
+There is a growing list of cloud native applications and services built
+using Rust. [AWS firecracker](#aws-fire-cracker) and [Linkerd service mesh
+proxy](#linkerd) are two examples. More projects could be found at [Rust
+Cloud Native GitHub](#rust-cloud-native-github).
+
+This paper describes how Rust applications can be integrated with the
+Cloud Native Data Plane ([CNDP](#cndp-io)) to rapidly develop flexible and
+high-performance packet processing applications.
+
+CNDP is a set of user space libraries created to address the need for a
+data plane framework designed for packet processing microservices in
+cloud deployments. CNDP aims to provide better performance than standard
+network socket interfaces using an I/O layer primarily built on Linux
+[AF_XDP](#af-xdp), an address family that is optimized for high
+performance packet processing. It provides an interface that delivers
+packets directly to user space, bypassing the kernel networking stack.
+More details on CNDP can be found at [cndp.io](#cndp-io)
+
+## CNDP Rust Crate 
+
+CNDP is natively implemented in the C programming language. CNDP
+provides language bindings for Rust which enable cloud native developers
+to build user space network functions using Rust and CNDP. The CNDP Rust
+crate provides high-level APIs that allow a Rust developer to interact
+with CNDP without having to fully understand the low-level details. It
+abstracts the complexities of network I/O and memory buffer management,
+allowing the developer to focus on their application.
+
+Rust interfaces with C with minimal overhead using a Foreign Function
+Interface (FFI) to C libraries. Rust provides two modules *std::ffi* and
+*std::os::raw* for transforming data types between Rust and C. In
+addition, Rust has tools like [bindgen](#rust-bindgen) which
+automatically generate Rust FFI bindings to C.
+
+CNDP provides two Rust crates
+
+1.  Low-level bindings crate (cne-sys) which exposes a minimal low-level
+    C interface to Rust.
+
+2.  High-level Rust APIs crate (cne) which uses low-level bindings
+    crate.
+
+Since CNDP is built on AF_XDP, it delivers packets directly to user
+space, bypassing the kernel networking stack. Rust user space network
+functions using the CNDP CNE crate can benefit from the network I/O
+performance offered by Linux kernel AF_XDP. It's expected that most Rust
+applications using CNDP only require the high-level Rust API crate and
+should not program directly to low-level crate.
+
+<div class="text-center">
+ <figure class="figure">
+  <img src="/assets/images/blog/CNDP-Rust-Crate.png" class="figure-img img-fluid" alt="CNDP Rust Crate">
+  <figcaption class="figure-caption text-center">Figure 1: CNDP Rust Crate</figcaption>
+ </figure>
+</div>
+
+
+(Figure 1) shows how the Rust application relates to CNDP and its Rust

```

Comment:
Can you remove the parenthesis? You can just say "Figure 1 shows how the Rust ..."

Reply:
ok. I saw the same parenthesis in CNDP blog when referring to Figure 1. Will remove it here